File: processing/syntactic/expressions/unary_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.operator import OperatorToken, InputSide
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.groups import InfixExpression, StatementExpression


class UnaryExpr(InfixExpression, StatementExpression):
    """ unary_expr:
        UNARY_LEFT prefix_expr
        | suffix_expr UNARY_RIGHT;
    """

    @child_property
    def value(self) -> Expr: pass

    def __init__(
            self,
            parent: Expr = None,
            operator: OperatorToken = None,
            value: Expr = None
    ):
        super().__init__(parent)
        self.operator = operator
        self.value = value
        # TODO: mark span depending on input side

    def to_axion(self, c: CodeBuilder):
        if self.operator.input_side == InputSide.left:
            c += self.value, ' ', self.operator
        elif self.operator.input_side == InputSide.right:
            c += self.operator, ' ', self.value
        else:
            raise NotImplementedError

    def to_csharp(self, c: CodeBuilder):
        self.to_axion(c)

    def to_python(self, c: CodeBuilder):
        self.to_axion(c)

File: Axion.Core/Processing/Syntactic/Expressions/Operations/UnaryExpr.cs
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Magnolia.Attributes;

namespace Axion.Core.Processing.Syntactic.Expressions.Operations;

/// <summary>
///     <code>
///         unary-expr:
///             UNARY-LEFT prefix-expr
///             | suffix-expr UNARY-RIGHT;
///     </code>
/// </summary>
[Branch]
public partial class UnaryExpr : PostfixExpr {
    [Leaf] OperatorToken @operator = null!;
    [Leaf] Node value = null!;

    public UnaryExpr(Node? parent) : base(parent) { }
}
